raiden/transfer/mediated_transfer/initiator.py

```diff
@@ -112,18 +128,23 @@ def try_new_route(state):
         # No available route has sufficient balance for the current transfer,
         # cancel it.
         #
-        # At this point we can just discard all the state data, this is only
-        # valid because we are the initiator and we know that the secret was
-        # not released.
+        # At this point we can just discard all the state data except for the
+        # reveal secret.
         transfer_failed = EventTransferSentFailed(
             identifier=state.transfer.identifier,
             reason='no route available',
         )
 
+        state.transfer.secret = None
+        state.transfer.hashlock = None
+        state.message = None
+        state.route = None
```

Comment:
This is discarding the route, so `handle_secretreveal` won't send the balance proof to the refund path, and the channel will eventually be settled on-chain.

Perhaps the best approach is to start a single task for each transfer, and have a retry state machine on top.

Reply:
I think I should clarify what I mean.

ATM (as of 3cb5d526f511707d4b4d0fa2840aa60d1cc861c6), the initiator task is conceptually managing from `1` transfer up to `len(state.routes.available_routes)` transfers. To make matters simpler, the current implementation cancels the latest transfer and starts a new one. That allows us to reuse the state object and logic from the same module.

The goal #1091 is to allow *multiple* transfers to happen simultaneously, we can imagine a worst case were more than one of the transfers fail with a refund. If the refunds are valid we want to reveal the secret to improve liquidity (we probably want to pay zero fees for this, but lets neglect this for now), this is in interest of the payer since it involves liquidity of **two** channels that he is participating.

To allow multiple transfers, we need to keep a list of the current pending transfers. This means for each transfer the attributes `transfer`, `message`, `route`, `secretrequest`, and `revealsecret` from the [initiator](https://github.com/raiden-network/raiden/blob/3cb5d526f511707d4b4d0fa2840aa60d1cc861c6/raiden/transfer/mediated_transfer/state.py#L47-L57) must be saved.

My proposal is to make a new state object that contains these fields, this state object will represent a single transfer started by the initiator, all logic to interact with this state object should be isolated into a single module, and then the current InitiatorState would sub-dispatch the state changes to the appropriate transfer task. When a refund is received for the current active transfer a new one should be started *along side* the current one, this allows the protocol to complete for the one that received the refund.

The proposed change allows multiple in-transit transfers, so that we can reveal the secret individually, and it also sets up the road for a better [cancel route](https://github.com/raiden-network/raiden/blob/3cb5d526f511707d4b4d0fa2840aa60d1cc861c6/raiden/transfer/mediated_transfer/initiator.py#L205). The current implementation  is sub-optimal because it just drops the current transfer and start a new one, instead it should let both transfers race and finish the first one that sends a secret request (IOW, we don't want to cancel a route, but to retry with a new route after a timeout).